linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -1491,15 +1491,17 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
             } else if (target.inlineable && (
                 target.shouldInline ||
                 shouldInlineBecauseOfArgs(target, treceiver :: targs))) {
-              inline(allocationSites, Some(treceiver), targs, target,
-                  isStat, usePreTransform)(cont)
+              val receiverType = ClassType(target.enclosingClassName)
+              inline(allocationSites, Some((receiverType, treceiver)), targs,
+                  target, isStat, usePreTransform)(cont)
             } else {
               treeNotInlined
             }
           } else {
             if (canMultiInline(impls)) {
-              inline(allocationSites, Some(treceiver), targs, impls.head,
-                  isStat, usePreTransform)(cont)
+              val receiverType = treceiver.tpe.base
```

Comment:
It feels odd that we are throwing the refined type away here. Would there be value in keeping it?

Reply:
We're not throwing it away; it's still present in `treceiver.tpe`, which will make its way to `withNewLocalDef` as `value.tpe`. We just need an accurate declared type, which is always a base `Type`. And here we cannot use `impls.head.enclosingClassName`, because we're multi-inlining, so in fact there the receiver is not generally of the target method's enclosing class type. For example:
```scala
class A

trait T {
  def foo(x: Int): Int = x + 1
}

class B extends A with T {
  def foo(x: Int): Int = super[T].foo(x)
}

class C extends A with T {
  def foo(x: Int): Int = super[T].foo(x)
}

def bar(t: T, x: Int): Int =
  t.foo(x)
```
In the last line, we can multi-inline `B.foo` and `C.foo` because they do the same thing. But we have to choose one, and `t` might be of the other's type.